node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeAllocation.java

```diff
@@ -290,17 +290,47 @@ public boolean fulfilledAndNoChanges() {
     }
 
     /**
-     * Returns {@link FlavorCount} describing the docker node deficit for the given {@link NodeSpec}.
+     * Returns {@link FlavorCount} describing the node deficit for the given {@link NodeSpec}.
      *
-     * @return empty if the requested spec is not count based or the requested flavor type is not docker or
-     *         the request is already fulfilled. Otherwise returns {@link FlavorCount} containing the required flavor
-     *         and node count to cover the deficit.
+     * @return empty if the requested spec is already fulfilled. Otherwise returns {@link FlavorCount} containing the
+     * flavor and node count required to cover the deficit.
      */
-    Optional<FlavorCount> getFulfilledDockerDeficit() {
-        return Optional.of(requestedNodes)
-                .filter(NodeSpec.CountNodeSpec.class::isInstance)
-                .map(spec -> new FlavorCount(spec.resources().get(), spec.fulfilledDeficitCount(accepted)))
-                .filter(flavorCount -> flavorCount.getCount() > 0);
+    Optional<FlavorCount> nodeDeficit() {
+        if (nodeType() != NodeType.config && nodeType() != NodeType.tenant) {
+            return Optional.empty(); // Requests for these node types never have a deficit
+        }
+        return Optional.of(new FlavorCount(requestedNodes.resources().orElseGet(NodeResources::unspecified),
+                                           requestedNodes.fulfilledDeficitCount(accepted())))
+                       .filter(flavorCount -> flavorCount.getCount() > 0);
+    }
+
+    /** Returns the indices to use when provisioning hosts for this */
+    List<Integer> provisionIndices(int count) {
+        if (count == 0) return List.of();
+        NodeType hostType = requestedNodes.type().hostType();
+
+        // Tenant hosts have a continuously increasing index
+        if (hostType == NodeType.host) return nodeRepository.database().readProvisionIndices(count);
+
+        // Infrastructure hosts have fixed indices: Their cluster index + 1
+        int offset = 1;
+        Set<Integer> currentIndices = allNodes.nodeType(hostType)
```

Comment:
It does look like the initial index maps to hostname index with the +1 relationship, but in zones that have replaced a cfg, their index is >= 3. We have no started filling index gaps, so I guess we will overtime return to all `config` being in [0-2], but isn't it possible if they get reprovisioned in "wrong" order, this breaks?
E.g.:
Say cfg1 has index 5, cfg2 has 6, cfg3 has 2. If we attempt to reprovision cfg2, it will be given index 1, which will overwrite the original cfg1's DNS records?

Reply:
Good point, we have to ensure current hosts have 0, 1, 2 or find another way to decide indices. Will look into it.